syntax = "proto3";

package stats;

option go_package = "./proto";

service StatsService {
  rpc GetPostStats (GetPostStatsRequest) returns (PostStatsResponse);
  rpc GetPostViewsDynamics (GetDynamicsRequest) returns (DynamicsResponse);
  rpc GetPostLikesDynamics (GetDynamicsRequest) returns (DynamicsResponse);
  rpc GetPostCommentsDynamics (GetDynamicsRequest) returns (DynamicsResponse);
  rpc GetTopPosts (GetTopRequest) returns (TopPostsResponse);
  rpc GetTopUsers (GetTopRequest) returns (TopUsersResponse);
}

message GetPostStatsRequest {
  string post_id = 1;
}

message PostStatsResponse {
  int64 views = 1;
  int64 likes = 2;
  int64 comments = 3;
}

message GetDynamicsRequest {
  string post_id = 1;
  string from_date = 2;
  string to_date = 3;
}

message DailyStat {
  string date = 1;
  int64 count = 2;
}

message DynamicsResponse {
  repeated DailyStat stats = 1;
}

message GetTopRequest {
  string stat_type = 1;
  int32 limit = 2;
}

message TopPost {
  string post_id = 1;
  int64 count = 2;
}

message TopPostsResponse {
  repeated TopPost posts = 1;
}

message TopUser {
  string user_id = 1;
  int64 count = 2;
}

message TopUsersResponse {
  repeated TopUser users = 1;
}